docs/Guidelines.md

```diff
@@ -0,0 +1,78 @@
+
+Dumping structs with `format.Object` is recommended. Starting with Kubernetes 1.26, format.Object will pretty-print Kubernetes API objects or structs as YAML and omit unset fields, which is more readable than other alternatives like `fmt.Sprintf("%+v")`.
+
+```golang
+import (
+    "fmt"
+    "k8s.io/api/core/v1"
+    "k8s.io/kubernetes/test/utils/format"
+)
+
+var pod v1.Pod
+fmt.Printf("format.Object:\n%s", format.Object(pod, 1 /* indent one level */))
+
+format.Object:
+    <v1.Pod>:
+        metadata:
+          creationTimestamp: null
+        spec:
+          containers: null
+        status: {}
+```
+## Polling and timeouts
+
+When waiting for something to happen, use a reasonable timeout. Without it, a test might keep running until the entire test suite gets killed by the CI. **Beware that the CI under load may take a lot longer to complete some operation compared to running the same test locally**. On the other hand, a too long timeout also has drawbacks:
+
+* When a feature is broken so that the expected state doesn’t get reached, a test waiting for that state first needs to time out before the test fails.
+* If a state is expected to be reached within a certain time frame, then a timeout that is much higher will cause test runs to be considered successful although the feature was too slow. A dedicated performance test in a well-know environment may be a better solution for testing such performance expectations.
+
+Good code that waits for something to happen meets the following criteria:
+* accepts a context for test timeouts
+* full explanation when it fails: when it observes some state and then encounters errors reading the state, then dumping both the latest observed state and the latest error is useful
+* early abort when condition cannot be reached anymore
+
+### Tips for writing and debugging long-running tests
+
+* Use `ginkgo.By` to record individual steps. Ginkgo will use that information when describing where a test timed out.
+* Use [gomega.Eventually](https://) to wait for some condition. When it times out or gets stuck, the last failed assertion will be included in the report automatically. A good way to invoke it is:
+```go
+	// Before injecting the token, validate that the linkedaccesstoken resource exists, otherwise injecting will return a 404 error code
+	Eventually(func() bool {
+		// application info should be stored even after deleting the application in application variable
+		_, err := s.GetSPIAccessToken(linkedAccessTokenName, namespace)
+		return err == nil
+	}, 1*time.Minute, 100*time.Millisecond).Should(BeTrue(), "SPI controller didn't create the SPIAccessToken")
```

Comment:
to be aligned with what was written in the previous section of "Debuggability" I would change the return type of a func that is passed into `Eventually`:

```suggestion
	Eventually(func() error {
		// application info should be stored even after deleting the application in application variable
		_, err := s.GetSPIAccessToken(linkedAccessTokenName, namespace)
		return err
	}, 1*time.Minute, 100*time.Millisecond).Should(Succeed(), "SPI controller didn't create the SPIAccessToken")
```
in this case, if Eventually will timeout, it will print out the value of the `error`, instead of "expected \<bool\> to be true", which doesn't hold much information

and I think it's worth to point out in the guide that the return type of functions passed to `Eventually` should always be something that tells us as much as information about what went wrong